File: ios/Yomiyasu/Core/Models/LibraryQueries.swift
import Foundation

struct SeriesQuery: Sendable, Equatable {
    var variant: LibraryVariant = .all
    var name: String?
    var genre: String?
    var author: String?
    var sort: SortValue = .seriesDefault
    var status: SerieStatus?
    var firstLetter: String?
    var minDifficulty: Int?
    var maxDifficulty: Int?
    var minValoration: Int?
    var maxValoration: Int?
    var valorationCount: Int?
    var readprogress: ProgressFilter?
    var readlistOnly: Bool = false
    var page: Int = 1
    var limit: Int = 25

    var isFiltering: Bool {
        genre != nil || author != nil || status != nil || minDifficulty != nil
            || maxDifficulty != nil || (readprogress != nil && readprogress != .all)
            || readlistOnly || firstLetter != nil
            || minValoration != nil || maxValoration != nil || valorationCount != nil
    }

    static func resetFilters(_ query: SeriesQuery) -> SeriesQuery {
        var result = query
        result.genre = nil
        result.author = nil
        result.status = nil
        result.firstLetter = nil
        result.minDifficulty = nil
        result.maxDifficulty = nil
        result.minValoration = nil
        result.maxValoration = nil
        result.valorationCount = nil
        result.readprogress = nil
        result.readlistOnly = false
        return result
    }

    var queryItems: [URLQueryItem] {
        var items: [URLQueryItem] = []

        if let name, !name.isEmpty {
            items.append(URLQueryItem(name: "name", value: name))
        }
        if let genre {
            items.append(URLQueryItem(name: "genre", value: genre))
        }
        if let author {
            items.append(URLQueryItem(name: "author", value: author))
        }
        items.append(URLQueryItem(name: "sort", value: sort.rawValue))
        if let status {
            items.append(URLQueryItem(name: "status", value: status.rawValue))
        }
        if let firstLetter {
            items.append(URLQueryItem(name: "firstLetter", value: firstLetter))
        }
        if let minDifficulty {
            items.append(URLQueryItem(name: "min", value: String(minDifficulty)))
        }
        if let maxDifficulty {
            items.append(URLQueryItem(name: "max", value: String(maxDifficulty)))
        }
        if let minValoration {
            items.append(URLQueryItem(name: "valorationMin", value: String(minValoration)))
        }
        if let maxValoration {
            items.append(URLQueryItem(name: "valorationMax", value: String(maxValoration)))
        }
        if let valorationCount {
            items.append(URLQueryItem(name: "valorationCount", value: String(valorationCount)))
        }
        if let readprogress, readprogress != .all {
            items.append(URLQueryItem(name: "readprogress", value: readprogress.rawValue))
        }
        if readlistOnly {
            items.append(URLQueryItem(name: "readlist", value: "true"))
        }

        items.append(URLQueryItem(name: "page", value: String(page)))
        items.append(URLQueryItem(name: "limit", value: String(limit)))

        return items
    }

    var alphabetQueryItems: [URLQueryItem] {
        var items: [URLQueryItem] = []

        if let genre {
            items.append(URLQueryItem(name: "genre", value: genre))
        }
        if let author {
            items.append(URLQueryItem(name: "author", value: author))
        }
        if let status {
            items.append(URLQueryItem(name: "status", value: status.rawValue))
        }
        if let minDifficulty {
            items.append(URLQueryItem(name: "min", value: String(minDifficulty)))
        }
        if let maxDifficulty {
            items.append(URLQueryItem(name: "max", value: String(maxDifficulty)))
        }
        if let minValoration {
            items.append(URLQueryItem(name: "valorationMin", value: String(minValoration)))
        }
        if let maxValoration {
            items.append(URLQueryItem(name: "valorationMax", value: String(maxValoration)))
        }
        if let valorationCount {
            items.append(URLQueryItem(name: "valorationCount", value: String(valorationCount)))
        }

        return items
    }

    var randomQueryItems: [URLQueryItem] {
        var items = alphabetQueryItems.filter { !Self.valorationParamNames.contains($0.name) }

        if let readprogress, readprogress != .all {
            items.append(URLQueryItem(name: "readprogress", value: readprogress.rawValue))
        }
        if readlistOnly {
            items.append(URLQueryItem(name: "readlist", value: "true"))
        }

        return items
    }

    private static let valorationParamNames: Set<String> = [
        "valorationMin", "valorationMax", "valorationCount",
    ]
}

struct BooksQuery: Sendable, Equatable {
    var variant: LibraryVariant = .all
    var name: String?
    var serie: String?
    var sort: SortValue?
    var status: ProgressStatus?
    var page: Int?
    var limit: Int?

    var queryItems: [URLQueryItem] {
        var items: [URLQueryItem] = []

        if let name, !name.isEmpty {
            items.append(URLQueryItem(name: "name", value: name))
        }
        if let serie {
            items.append(URLQueryItem(name: "serie", value: serie))
        }
        if let sort {
            items.append(URLQueryItem(name: "sort", value: sort.rawValue))
        }
        if let status {
            items.append(URLQueryItem(name: "status", value: status.rawValue))
        }
        if let page {
            items.append(URLQueryItem(name: "page", value: String(page)))
        }
        if let limit {
            items.append(URLQueryItem(name: "limit", value: String(limit)))
        }

        return items
    }
}

File: android/app/src/main/java/es/manabe/yomiyasu/core/models/LibraryQueries.kt
package es.manabe.yomiyasu.core.models

data class SeriesQuery(
    val variant: LibraryVariant = LibraryVariant.All,
    val name: String? = null,
    val genre: String? = null,
    val author: String? = null,
    val sort: SortValue = SortValue.SeriesDefault,
    val status: SerieStatus? = null,
    val firstLetter: String? = null,
    val minDifficulty: Int? = null,
    val maxDifficulty: Int? = null,
    val minValoration: Int? = null,
    val maxValoration: Int? = null,
    val valorationCount: Int? = null,
    val readprogress: ProgressFilter? = null,
    val readlistOnly: Boolean = false,
    val page: Int = 1,
    val limit: Int = 25,
) {
    val isFiltering: Boolean
        get() = genre != null || author != null || status != null || minDifficulty != null ||
            maxDifficulty != null || (readprogress != null && readprogress != ProgressFilter.All) ||
            readlistOnly || firstLetter != null ||
            minValoration != null || maxValoration != null || valorationCount != null

    fun resetFilters(): SeriesQuery = copy(
        genre = null,
        author = null,
        status = null,
        firstLetter = null,
        minDifficulty = null,
        maxDifficulty = null,
        minValoration = null,
        maxValoration = null,
        valorationCount = null,
        readprogress = null,
        readlistOnly = false,
    )

    val queryItems: List<Pair<String, String>>
        get() = buildList {
            name?.takeIf { it.isNotEmpty() }?.let { add("name" to it) }
            genre?.let { add("genre" to it) }
            author?.let { add("author" to it) }
            add("sort" to sort.rawValue)
            status?.let { add("status" to it.rawValue) }
            firstLetter?.let { add("firstLetter" to it) }
            minDifficulty?.let { add("min" to it.toString()) }
            maxDifficulty?.let { add("max" to it.toString()) }
            minValoration?.let { add("valorationMin" to it.toString()) }
            maxValoration?.let { add("valorationMax" to it.toString()) }
            valorationCount?.let { add("valorationCount" to it.toString()) }
            readprogress?.takeIf { it != ProgressFilter.All && it.rawValue != null }
                ?.let { add("readprogress" to it.rawValue!!) }
            if (readlistOnly) add("readlist" to "true")
            add("page" to page.toString())
            add("limit" to limit.toString())
        }

    val alphabetQueryItems: List<Pair<String, String>>
        get() = buildList {
            genre?.let { add("genre" to it) }
            author?.let { add("author" to it) }
            status?.let { add("status" to it.rawValue) }
            minDifficulty?.let { add("min" to it.toString()) }
            maxDifficulty?.let { add("max" to it.toString()) }
            minValoration?.let { add("valorationMin" to it.toString()) }
            maxValoration?.let { add("valorationMax" to it.toString()) }
            valorationCount?.let { add("valorationCount" to it.toString()) }
        }

    val randomQueryItems: List<Pair<String, String>>
        get() = buildList {
            addAll(alphabetQueryItems.filterNot { it.first in valorationParamNames })
            readprogress?.takeIf { it != ProgressFilter.All && it.rawValue != null }
                ?.let { add("readprogress" to it.rawValue!!) }
            if (readlistOnly) add("readlist" to "true")
        }

    private companion object {
        val valorationParamNames = setOf("valorationMin", "valorationMax", "valorationCount")
    }
}

data class BooksQuery(
    val variant: LibraryVariant = LibraryVariant.All,
    val name: String? = null,
    val serie: String? = null,
    val sort: SortValue? = null,
    val status: ProgressStatus? = null,
    val page: Int? = null,
    val limit: Int? = null,
) {
    val queryItems: List<Pair<String, String>>
        get() = buildList {
            name?.takeIf { it.isNotEmpty() }?.let { add("name" to it) }
            serie?.let { add("serie" to it) }
            sort?.let { add("sort" to it.rawValue) }
            status?.let { add("status" to it.key) }
            page?.let { add("page" to it.toString()) }
            limit?.let { add("limit" to it.toString()) }
        }
}
